src/contents/l1sload-guide-read-the-l1-state-from-l2.mdx

```diff
@@ -0,0 +1,346 @@
+    }
+}`} language="solidity"/>
+
+## Example #3: Reading ERC20 token balance from L1
+
+Let's start by deploying the following very simple ERC20 token.
+
+<ContentCopy code={`// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.17;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+
+// In our final example, we'll read the balance of any ERC20 holder on L1
+// Please note that we will be using OpenZeppelin's implementation which puts the balance mapping on slot 0, is is not enforced by the ERC20 standard
+contract SimpleToken is ERC20 {
+    constructor() ERC20("Simple Token", "STKN") {
+        _mint(msg.sender, 21_000_000 ether);
+    }
+}`} language="solidity"/>
+
+Next, we can deploy the following contract on L2 by passing the L1 token address as parameter.
+
+<ContentCopy code={`// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity ^0.8.20;
+
+interface IL1Blocks {
+    function latestBlockNumber() external view returns (uint256);
+}
+
+// This contract reads the balance of any holder on L1
+contract L2Storage {
+    address constant L1_BLOCKS_ADDRESS = 0x5300000000000000000000000000000000000001;
+    address constant L1_SLOAD_ADDRESS = 0x0000000000000000000000000000000000000101;
+    address immutable l1TokenAddress;
+
+    constructor(address _l1TokenAddress) {
+        l1TokenAddress = _l1TokenAddress;
+    }
+
+    // Retrieves the token balance of a given Ethereum account.
+    function retrieveL1Balance(address account) public view returns(uint) {
+        // We assume that the balance mapping is stored at slot number 0
+        uint slotNumber = 0;
+        // The formula that Solidity uses to compute the slot in mappings is: balanceSlotPosition = keccak256(holderAddress, slotNumber)
+        uint accountBalanceSlot = uint(
+            keccak256(abi.encodePacked(uint(uint160(account)),
+            slotNumber)
+        ));
+        // Now, we perform a staticcall to the l1sload precompile to retrieve the account balance
+        bool success;
+        bytes memory returnValue;
+        (success, returnValue) = L1_SLOAD_ADDRESS.staticcall(abi.encodePacked(l1TokenAddress, accountBalanceSlot));
+        if(!success)
+        {
+            revert("L1SLOAD failed");
+        }
+        // The retrieved value is in bytes32 format, so we cast it to uint256 before returning it
+        return abi.decode(returnValue, (uint));
+    }
+}`} language="solidity"/>
+
+OpenZeppelin contracts conveniently places [the balances mapping on Slot 0](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/ERC20.sol#L30). So you can call `retrieveL1Balance()` by passing the account address as paramater and the token balance will be stored on the `l1Balance` variable. As you can see on the code, it works by converting the account to `uint160` and then hashing it with the mapping slot which is 0. This is because that's the way the Solidity implemented mappings.
```

Comment:
This could just be me missing this - do you know where we can reference the `l1Balance` variable in the example?